src/datatablerow/datatablerow.jl

```diff
@@ -37,31 +35,119 @@ Base.done(r::DataTableRow, s) = s > length(r)
+Base.@propagate_inbounds hash_colel{T}(v::NullableArray{T}, i, h::UInt = zero(UInt)) =
+    isnull(v, i) ? hash(NULL_MAGIC, h) : hash(get(v[i]), h)
+Base.@propagate_inbounds hash_colel{T}(v::AbstractCategoricalArray{T}, i, h::UInt = zero(UInt)) =
+    hash(CategoricalArrays.index(v.pool)[v.refs[i]], h)
+Base.@propagate_inbounds function hash_colel{T}(v::AbstractCategoricalArray{T}, i, h::UInt = zero(UInt))
+    ref = v.refs[i]
+    ref == 0 ? hash(NULL_MAGIC, h) : hash(CategoricalArrays.index(v.pool)[ref], h)
+end
+
 # hash of DataTable rows based on its values
 # so that duplicate rows would have the same hash
-function Base.hash(r::DataTableRow, h::UInt)
-    for col in columns(r.dt)
-        if _isnull(col[r.row])
-            h = hash(false, h)
-        else
-            h = hash(true, hash(col[r.row], h))
-        end
+function rowhash(dt::DataTable, r::Int, h::UInt = zero(UInt))
+    @inbounds for col in columns(dt)
+        h = hash_colel(col, r, h)
     end
     return h
 end
 
+Base.hash(r::DataTableRow, h::UInt = zero(UInt)) = rowhash(r.dt, r.row, h)
+
 # comparison of DataTable rows
-# only the rows of the same DataTable could be compared
 # rows are equal if they have the same values (while the row indices could differ)
-@compat(Base.:(==))(r1::DataTableRow, r2::DataTableRow) = isequal(r1, r2)
-
-function Base.isequal(r1::DataTableRow, r2::DataTableRow)
-    r1.dt == r2.dt || throw(ArgumentError("Comparing rows from different frames not supported"))
-    r1.row == r2.row && return true
-    for col in columns(r1.dt)
-        if !isequal(col[r1.row], col[r2.row])
-            return false
+# returns Nullable{Bool}
+# if all non-null values are equal, but there are nulls, returns null
+function @compat(Base.:(==))(r1::DataTableRow, r2::DataTableRow)
+    if r1.dt !== r2.dt
+        (ncol(r1.dt) != ncol(r2.dt)) &&
+            throw(ArgumentError("Comparing rows from different frames not supported"))
+        eq = Nullable(true)
+        @inbounds for (col1, col2) in zip(columns(r1.dt), columns(r2.dt))
+            eq_col = convert(Nullable{Bool}, col1[r1.row] == col2[r2.row])
+            # If true or null, need to compare remaining columns
+            get(eq_col, true) || return Nullable(false)
+            eq &= eq_col
         end
+        return eq
+    else
+    	r1.row == r2.row && return Nullable(true)
+        eq = Nullable(true)
+        @inbounds for col in columns(r1.dt)
+            eq_col = convert(Nullable{Bool}, col[r1.row] == col[r2.row])
+            # If true or null, need to compare remaining columns
+            get(eq_col, true) || return Nullable(false)
+            eq &= eq_col
+        end
+        return eq
+    end
+end
+
+# internal method for comparing the elements of the same data frame column
+isequal_colel(col::AbstractArray, r1::Int, r2::Int) =
+    (r1 == r2) || isequal(Base.unsafe_getindex(col, r1), Base.unsafe_getindex(col, r2))
```

Comment:
`@inbounds` would not work, because AFAIR the macro interfers with the result of an expression. `@propagate_inbounds` is much nicer anyway, I became aware of it too late :)